### PR TITLE
perf: add type-specialized fast paths for collection marshal (variable-length lists + maps)

### PR DESCRIPTION
**Pull request by mykaul**
## Summary

Add concrete type-switch fast paths to `marshalList` and `marshalMap` for common variable-length list types and all map types. Complements PR #839 (fixed-size list marshal) and PR #881 (unmarshalList fast paths).

### Supported types

**Variable-length list marshal fast paths** (6 types):

| Go type | CQL types | Wire strategy |
|---------|-----------|---------------|
| `[]string` | varchar, text, ascii | Two-pass: accumulate sizes → single alloc + copy |
| `[]bool` | boolean | Single alloc: `4 + n*(4+1)` |
| `[][]byte` | blob | Two-pass + null element support |
| `[]int16` | smallint | Single alloc: `4 + n*(4+2)` |
| `[]int8` | tinyint | Single alloc: `4 + n*(4+1)` |
| `[]UUID` | uuid, timeuuid | Single alloc: `4 + n*(4+16)` |

**Map marshal fast paths** (11 types):

| Go type | CQL key→val | Wire strategy |
|---------|-------------|---------------|
| `map[string]string` | text→text | Two-pass (accumulate key+val sizes) |
| `map[string]int64` | text→bigint | Accumulate key sizes + fixed val |
| `map[string]int` | text→int | Accumulate key sizes + fixed val |
| `map[string]float64` | text→double | Accumulate key sizes + fixed val |
| `map[string]float32` | text→float | Accumulate key sizes + fixed val |
| `map[string]bool` | text→boolean | Accumulate key sizes + fixed val |
| `map[string][]byte` | text→blob | Two-pass + null element support |
| `map[string]int16` | text→smallint | Accumulate key sizes + fixed val |
| `map[int64]string` | bigint→text | Accumulate val sizes + fixed key |
| `map[int64]int64` | bigint→bigint | Single alloc, single-pass |
| `map[int64]float64` | bigint→double | Single alloc, single-pass |

### Key design decisions

- **Type-switch before reflect** — exact Go type match; non-matching types fall through to the unchanged generic path.
- **Single output allocation** — every fast path computes total wire size upfront and does one `make([]byte, total)`.
- **No per-element Marshal calls** — eliminates the per-element `[]byte(str)` heap alloc that dominates the generic path.
- **Consistent with #839/#881 patterns** — same type-switch placement, same CQL type validation, same overflow guards.
- **Null element support** — `nil` `[]byte` elements are encoded as CQL null (length -1), not empty (length 0).

### Benchmark results

`go test -bench=BenchmarkCompareMarshal -benchtime=500ms -count=1 -tags unit`
CPU: 12th Gen Intel(R) Core(TM) i7-1270P

"before" = named/alias types that bypass the type-switch and fall through to the unchanged generic reflect path (master behavior). "after" = this PR.

#### Lists — n=100 (realistic/medium collection)

| Type | before (ns/op) | after (ns/op) | speedup | before allocs | after allocs |
|------|---------------|--------------|---------|--------------|-------------|
| `[]string` | 7,267 | 674 | **10.8x** | 206 | 1 |
| `[]bool` | 4,301 | 232 | **18.5x** | 204 | 1 |
| `[][]byte` | 6,105 | 734 | **8.3x** | 106 | 1 |
| `[]int16` | 4,735 | 273 | **17.3x** | 205 | 1 |
| `[]int8` | 4,164 | 294 | **14.2x** | 204 | 1 |
| `[]UUID` | 8,552 | 896 | **9.5x** | 306 | 1 |

#### Maps — n=100 (realistic/medium collection)

| Type | before (ns/op) | after (ns/op) | speedup | before allocs | after allocs |
|------|---------------|--------------|---------|--------------|-------------|
| `map[string]string` | 16,265 | 3,669 | **4.4x** | 408 | 3 |
| `map[string]int64` | 16,046 | 2,311 | **6.9x** | 408 | 3 |
| `map[string]int` | 15,436 | 2,657 | **5.8x** | 407 | 3 |
| `map[string]float64` | 18,465 | 2,801 | **6.6x** | 408 | 3 |
| `map[string]bool` | 14,582 | 2,185 | **6.7x** | 407 | 3 |
| `map[int64]int64` | 13,929 | 1,403 | **9.9x** | 408 | 1 |

#### Lists — n=10 (small collection)

| Type | before (ns/op) | after (ns/op) | speedup | before allocs | after allocs |
|------|---------------|--------------|---------|--------------|-------------|
| `[]string` | 736 | 84 | **8.8x** | 22 | 1 |
| `[]bool` | 443 | 45 | **9.9x** | 21 | 1 |
| `[][]byte` | 609 | 91 | **6.7x** | 12 | 1 |
| `[]int16` | 452 | 42 | **10.6x** | 21 | 1 |
| `[]int8` | 437 | 45 | **9.8x** | 21 | 1 |
| `[]UUID` | 957 | 104 | **9.2x** | 33 | 1 |

#### Maps — n=10 (small collection)

| Type | before (ns/op) | after (ns/op) | speedup | before allocs | after allocs |
|------|---------------|--------------|---------|--------------|-------------|
| `map[string]string` | 1,672 | 371 | **4.5x** | 44 | 3 |
| `map[string]int64` | 1,561 | 345 | **4.5x** | 44 | 3 |
| `map[string]int` | 1,544 | 323 | **4.8x** | 44 | 3 |
| `map[string]float64` | 1,592 | 315 | **5.0x** | 44 | 3 |
| `map[string]bool` | 1,511 | 272 | **5.6x** | 44 | 2 |
| `map[int64]int64` | 1,443 | 195 | **7.4x** | 44 | 1 |

#### Lists — n=1000 (large/exceptional)

| Type | before (ns/op) | after (ns/op) | speedup | before allocs | after allocs |
|------|---------------|--------------|---------|--------------|-------------|
| `[]string` | 72,268 | 7,416 | **9.7x** | 2,009 | 1 |
| `[]bool` | 42,610 | 2,874 | **14.8x** | 2,008 | 1 |
| `[]int16` | 43,454 | 2,821 | **15.4x** | 2,008 | 1 |
| `map[string]string` | 171,752 | 32,085 | **5.4x** | 4,011 | 3 |
| `map[int64]int64` | 131,549 | 14,010 | **9.4x** | 4,011 | 1 |

### Testing

- **Differential byte-equality tests**: every list fast path is verified byte-for-byte against the generic reflect path (via named/alias types) for each supported type, including nil, empty, negative, zero, and boundary values.
- **Map round-trip tests**: map differential tests unmarshal both outputs and compare via `reflect.DeepEqual` to handle iteration order nondeterminism.
- **Full test suite**: 10,680+ tests pass with `-race -tags unit`. No regressions.

### CODE REVIEW

File: marshal_test.go [hunk 1]
```diff
@@ -172,6 +172,411 @@ func equalStringPointerSlice(leftList, rightList []*string) bool {
...
+type namedMapStringInt64     map[string]int64
+type namedMapStringInt       map[string]int
+type namedMapStringFloat64   map[string]float64
+type namedMapStringFloat32   map[string]float32
+type namedMapStringBool      map[string]bool
+type namedMapStringBytes     map[string][]byte
+type namedMapStringInt16     map[string]int16
+type namedMapInt64String     map[int64]string
+type namedMapInt64Int64      map[int64]int64
+type namedMapInt64Float64    map[int64]float64
```
[the remark this answers is not in the document]
Comment (mykaul) [reply marshal_test.go]: Done

File: marshal.go [hunk 2]
```diff
@@ -2202,3 +2340,466 @@ func wrapUnmarshalError(err error, msg string) UnmarshalError {
...
+	off := 4
+	for i := range keys {
+		binary.BigEndian.PutUint32(buf[off:], uint32(len(keys[i])))
+		off += 4
+		copy(buf[off:], keys[i])
+		off += len(keys[i])
+		binary.BigEndian.PutUint32(buf[off:], uint32(len(vals[i])))
+		off += 4
+		copy(buf[off:], vals[i])
+		off += len(vals[i])
```
[the remark this answers is not in the document]
Comment (mykaul) [reply marshal.go]: Done

**Comment by mykaul** [comment]
Fixed in 760f7957. Changed the overflow test value to `int(int64(math.MaxInt32) + 1)` so it compiles on 32-bit architectures while still producing the overflow at runtime on 64-bit.

**Comment by mykaul** [comment]
Closing as duplicate of #910
